fs-index/src/watch.rs

```diff
@@ -0,0 +1,157 @@
+pub async fn watch_index<P: AsRef<Path>, Id: ResourceId>(
+    root_path: P,
+) -> Result<()> {
+    log::debug!(
+        "Attempting to watch index at root path: {:?}",
+        root_path.as_ref()
+    );
+
+    let root_path = fs::canonicalize(root_path.as_ref())?;
+    let mut index: ResourceIndex<Id> = ResourceIndex::build(&root_path)?;
+    index.store()?;
+
+    let (mut watcher, mut rx) = async_watcher()?;
+    info!("Watching directory: {:?}", root_path);
+    let config = Config::default();
+    watcher.configure(config)?;
+    watcher.watch(root_path.as_ref(), RecursiveMode::Recursive)?;
+    info!("Started watcher with config: \n\t{:?}", config);
+
+    let ark_folder = root_path.join(ARK_FOLDER);
+    while let Some(res) = rx.next().await {
+        match res {
+            Ok(event) => {
+                // If the event is a change in .ark folder, ignore it
+                if event
+                    .paths
+                    .iter()
+                    .any(|p| p.starts_with(&ark_folder))
+                {
+                    continue;
+                }
+                // We only care for:
+                // - file modifications
+                // - file renames
+                // - file creations
+                // - file deletions
+                match event.kind {
+                    notify::EventKind::Modify(_) => match event {
+                        notify::Event {
+                            kind:
+                                notify::EventKind::Modify(
+                                    notify::event::ModifyKind::Data(_),
+                                ),
+                            ..
+                        } => {}
+                        notify::Event {
+                            kind:
+                                notify::EventKind::Modify(
+                                    notify::event::ModifyKind::Name(_),
+                                ),
+                            ..
+                        } => {}
+                        _ => continue,
+                    },
+                    notify::EventKind::Create(_) => match event {
+                        notify::Event {
+                            kind:
+                                notify::EventKind::Create(
+                                    notify::event::CreateKind::File,
+                                ),
+                            ..
+                        } => {}
+                        _ => continue,
+                    },
+                    notify::EventKind::Remove(_) => match event {
+                        notify::Event {
+                            kind:
+                                notify::EventKind::Remove(
+                                    notify::event::RemoveKind::File,
+                                ),
+                            ..
+                        } => {}
+                        _ => continue,
+                    },
+                    _ => {}
+                }
```

Comment:
Can't we match `event` directly without nested `match` expressions?

Reply:
Yes, I’m not sure why I wrote it that way in the first place 😃 

Updated